### PR TITLE
BQuest Complete

### PR DESCRIPTION
**Pull request by cbolles**
Transferring of settings from EEPROM to the BQ chip complete. This has the ability to transfer the settings one-by-one from EEPROM to the BQ and validate that the transfer worked successfully.

### CODE REVIEW

File: include/BMS/BQSettingStorage.hpp [hunk 1]
```diff
@@ -93,6 +92,12 @@ class BQSettingsStorage {
      */
     EVT::core::DEV::M24C32& getEEPROM();
 
+    /**
+     * Transfer settings from EEPROM to the BQ chip. This will read the
+     * settings one-by-one and pass the setting to the BQ chip for programming.
+     */
+    BMS::DEV::BQ76952::Status transferSettings();
```
Comment (mjmagee991) [inline include/BMS/BQSettingStorage.hpp:99]: Is there a reason this is in this file instead of the BQ76952 file?

File: src/dev/BQ76952.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    // Make sure the device actually exited Config Update Mode
+    bool isInConfigMode;
+    RETURN_IF_ERR(inConfigMode(&isInConfigMode));
+    if (isInConfigMode) {
+        return Status::ERROR;
+    }
+
+    return Status::OK;
+
+    return Status::OK;
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:65]: This line will never run.

File: src/dev/BQ76952.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    if (setting.getSettingType() != BMS::BQSetting::BQSettingType::RAM) {
+        return Status::ERROR;
+    }
+    return writeRAMSetting(setting);
+}
+
+BQ76952::Status BQ76952::enterConfigUpdateMode() {
+    // Number of times it will wait to see if the device has entered
+    // config update mode
+    // static constexpr uint8_t NUM_ATTEMPTS = 10;
```
Comment (mjmagee991) [inline src/dev/BQ76952.cpp:37]: Should this be deleted, or are you planning to implement this later?

File: targets/bq_interface/main.cpp [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+    uart.printf("Register 0x%x: 0x%08X\r\n", reg, ramValue);
+}
+
+/**
+ * Function for making a direct write request
+ *
+ * @param[in] uart The UART interface to write in from
+ */
+void directWrite(IO::UART& uart) {}
```
Comment (mjmagee991) [inline targets/bq_interface/main.cpp:103]: Did you mean to implement these functions?

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Nice work. That's a huge PR. I had a couple comments, but great job.

File: include/BMS/BMS.hpp [hunk 5]
```diff
@@ -1,6 +1,8 @@
-#include <stdint.h>
-#include <Canopen/co_core.h>
+#pragma once
```
[the remark this answers is not in the document]
Comment (cbolles) [reply include/BMS/BMS.hpp:1]: At one point I was dealing with something that I thought may have been an issue with an include guard. So to remove the chance of the issue being a spelling mistake in the include guard, I used `#pragma once` instead

File: include/BMS/BMS.hpp [hunk 6]
```diff
@@ -52,47 +59,37 @@ class BMS {
...
-            .Type = 0,
-            .Data = (uintptr_t)0x12
-        },
-        {
-            .Key = CO_KEY(0x1018, 4, CO_UNSIGNED32|CO_OBJ_D__R_),
+            .Key = CO_KEY(0x1018, 1, CO_UNSIGNED32 | CO_OBJ_D__R_),
             .Type = 0,
-            .Data = (uintptr_t)0x13
-        },
+            .Data = (uintptr_t) 0x10},
```
[the remark this answers is not in the document]
Comment (cbolles) [reply include/BMS/BMS.hpp:72]: Good tip! Thanks

File: include/BMS/dev/BQ76952.hpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+     * @param[in] reg The I2C register address to read from
+     * @param[out] result The data that was read
+     * @return The status of the read request attempt
+     */
+    Status makeDirectRead(uint8_t reg, uint16_t* result);
+
+    /**
+     * Execute a subcommand read request.
+     *
+     * Will make a subcommand read request to the BQ chup.
```
[the remark this answers is not in the document]
Comment (cbolles) [reply include/BMS/dev/BQ76952.hpp:86]: Thanks!

File: src/dev/BQ76952.cpp (hunk 2 above)
Comment (cbolles) [reply src/dev/BQ76952.cpp:65]: woops

File: targets/bq_interface/main.cpp (hunk 4 above)
Comment (cbolles) [reply targets/bq_interface/main.cpp:103]: Not at the moment, later this may be needed for testing, but for this functionality it wasn't needed

**Review by mjmagee991 — APPROVED** [review]
Nice work

**Pull request merged**